prototype/mobile_endpoint/dao.py

```diff
@@ -0,0 +1,238 @@
+        Commit the transaction
+        """
+        pass
+
+    @abstractmethod
+    def get_synclog(self, id):
+        pass
+
+    @abstractmethod
+    def get_form(self, id):
+        pass
+    
+    @abstractmethod
+    def get_case(self, id, lock=False):
+        pass
+
+    @abstractmethod
+    def case_exists(self, id):
+        pass
+
+    @abstractmethod
+    def iter_cases(self, case_ids, chunksize=100, ordered=True):
+        pass
+
+    @abstractmethod
+    def get_reverse_indexed_cases(self, domain, case_ids):
+        pass
+
+
+class SQLDao(AbsctractDao):
+    def commit(self, xform, cases):
+        def to_sql(doc):
+            if isinstance(doc, XFormInstance):
+                return FormData.from_generic(doc)
+            elif isinstance(doc, CommCareCase):
+                return CaseData.from_generic(doc)
+
+        def get_indices():
+            for case in cases:
+                for index in case.indices:
+                    yield CaseIndex.from_generic(index, case.id)
+
+        new_form, xform_sql = to_sql(xform)
+        case_docs = map(to_sql, cases)
+
+        indices = list(get_indices())
+
+        for new_case, case_sql in case_docs:
+            case_sql.forms.append(xform_sql)
+
+        combined = [(new_form, xform_sql)] + case_docs + indices
+        for is_new, doc in combined:
+            if is_new:
+                db.session.add(doc)
+
+        db.session.commit()
+
+    @to_generic
+    def get_synclog(self, id):
+        return Synclog.query.get(id)
+
+    @to_generic
+    def get_form(self, id):
+        return FormData.query.get(id)
+
+    @to_generic
+    def get_case(self, id, lock=False):
+        if lock:
+            return get_with_lock('case_lock_{}'.format(id), lambda: CaseData.query.get(id))
+        else:
+            None, CaseData.query.get(id)
+
+    def case_exists(self, id):
+        return CaseData.query.filter_by(id=id).exists()
+
+    @to_generic
+    def iter_cases(self, case_ids, chunksize=100, ordered=False):
+        for doc_ids in chunked(case_ids, chunksize):
+            cases = CaseData.query.filter(CaseData.id.in_(doc_ids))
+            if ordered:
+                # SQL won't return the rows in any particular order so we need to order them ourselves
+                index_map = {UUID(id_): index for index, id_ in enumerate(case_ids)}
+                ordered_cases = [None] * len(case_ids)
+                for case in cases:
+                    ordered_cases[index_map.pop(UUID(case.id))] = case
+
+                assert not index_map
+                cases = ordered_cases
+
+            for case in cases:
+                yield case
+
+    @to_generic
+    def get_reverse_indexed_cases(self, domain, case_ids):
+        return CaseData.query.join('indices')\
+            .filter(CaseIndex.referenced_id.in_(case_ids))\
+            .options(
+                contains_eager('indices'),
+                defer(CaseData.case_json)
+        )
+
+
+class CaseDbCache(object):
```

Comment:
(musing) i could be wrong, but it feels like a clean implementation shouldn't need this object. seems like we should just be able to get all the cases we know we might touch up front and save some complexity